apps/05Mastermind.js

```diff
@@ -9,47 +9,118 @@ var board = [];
 var solution = '';
 var letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
 
+
+
+
 function printBoard() {
     for (var i = 0; i < board.length; i++) {
         console.log(board[i])
     }
 }
 
+
+
 function generateSolution() {
     for (var i = 0; i < 4; i++) {
         var randomIndex = getRandomInt(0, letters.length);
         solution += letters[randomIndex];
     }
 }
 
+
 function getRandomInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min;
 }
 
+
+
+
+
+
+
+
+
+
+
+
+
+
+
 function generateHint(solution, guess) {
     // your code here
+    // generateHint() is a function that is created to do exactly
+    // what it says it will do, generate a hint.
+    // spec 2 is broken down into several parts
+    //
+    // the first thing we know is that it takes in two arguments
+    // the two arguements are 'solution' and 'guess'
+    // arguments are the named parameters placed in the parantheses immediately
+    // following the function name. - done
+    //
+    //
+    // 2.1
+    // create a variable named solutionArray
+    // create a variable named guessArray
     var solutionArray = solution.split('');
     var guessArray = guess.split('');
-    var correctLetterLocations = 0;
-    for (var i = 0; guessArray[i] === solutionArray[i]; i++) {
-      correctLetterLocations++;
-      solutionArray[i] = null;
+    var correctLetterLocations = 0; // will record how many correct "letter-locations" were guessed
+    var correctLetters = 0;
+
+
+    for (var i = 0; i < solutionArray.length; i++) {
+      if (guessArray[i] === solutionArray[i]) {
+        correctLetterLocations++;
+        solutionArray[i] = null;
+      }
+    }
+
+    for (var i = 0; i < solutionArray.length; i++) {
+      var targetIndex = guessArray.indexOf(solutionArray[i]);
+
+//the following if statement simply tells us what to do if
+// there is a matching item in guestArray and solutionArray
+// if it is greater than - 1 then we know something has been found as this
+// is what the indexOf method definition tells us //
+
+
+      if (targetIndex > - 1) {
+        correctLetters++;
+        solutionArray[i] = null;
+      }
     }
-  }
+
+
+    var hintString = correctLetterLocations + "-" + correctLetters;
+      return hintString;
+      console.log(hintString);
+
+
+}
+
+
+
+
+
 
 
 
 function mastermind(guess) {
     // your code here
-
-    solution = 'abcd';
+    if (guess >= 10) {
+      console.log('too many guesses!')
+    }
+    solution = 'abcd'; //Spec 0 - this is a global variable
+    var hint = generateHint(solution, guess);
 
     if (guess === solution) {
-      return ('You guessed it!')
-    }
+      return ('You guessed it!');
+    } // Spec 1 - this is what happens if you 'win' the game
+
+    if (guess !== solution) {
+      generateHint(solution, guess);
```

Comment:
Each time I read it, I understand a little more. Compared to 2 weeks ago I understand 1000000 percent more. 

I imagine in a few months I will be able to look back on this stuff and have a totally different way of thinking and looking at it. 

I'm starting to recognize patterns in applications, how they are laid out..

it seems like we always have 

1. a user prompt (duh, except fot that shitty game myst maybe)
2. global variables defined at the top that will be of major importance later on 
3. functions that perform the actions of the games
4. conditionals within those funcitons that present various scenarioes and results based on the data receieved 
5. Loops to make it easier to understand and less typing 
6. one main function (the game) that runs all of the functions. 
7. and alot of objects to make storing this stuff easier and for greater customization and a billonity other things.